src/box/alter.cc

```diff
@@ -1830,56 +1830,83 @@ alter_space_move_indexes(struct alter_space *alter, uint32_t begin,
 	return 0;
 }
 
+/**
+ * Helper of @sa update_view_references.
+ * Update view reference counter of a space by its given @a space_name,
+ * adding @a update_value to the counter.
+ * If @a suppress_error is true - always return 0; otherwise, and if the space
+ * was not found (or not applicable or etc) - diag is set and -1 is returned.
+ */
+static int
+update_view_space_reference(struct Select *select, const char *space_name,
+			    int update_value, bool suppress_error)
+{
+	if (space_name == NULL)
+		return 0;
+	/*
+	 * Views are allowed to contain CTEs. CTE is a temporary object,
+	 * created and destroyed at SQL runtime (it is represented by an
+	 * ephemeral table). So, it is absent in space cache and as a
+	 * consequence we can't increment its reference counter.
+	 * Skip iteration.
+	 */
+	if (sql_select_constains_cte(select, space_name))
+		return 0;
+	struct space *space = space_by_name0(space_name);
+	if (space == NULL) {
+		if (suppress_error)
+			return 0;
+		diag_set(ClientError, ER_NO_SUCH_SPACE, space_name);
+		return -1;
+	}
+	assert(space->def->view_ref_count > 0 || update_value > 0);
+	space->def->view_ref_count += update_value;
+	return 0;
+}
+
 /**
  * Walk through all spaces from 'FROM' clause of given select,
  * and update their view reference counters.
  *
  * @param select Tables from this select to be updated.
  * @param update_value +1 on view creation, -1 on drop.
- * @param suppress_error If true, silently skip nonexistent
+ * @param suppress_error If true - silently skip nonexistent (or non-applicable)
  *                       spaces from 'FROM' clause.
- * @param[out] not_found_space Name of a disappeared space.
- * @retval 0 on success, -1 if suppress_error is false and space
- *         from 'FROM' clause doesn't exist.
+ * @retval 0 on success or if suppress_error is true; -1 otherwise, diag is set.
  */
 static int
-update_view_references(struct Select *select, int update_value,
-		       bool suppress_error, const char **not_found_space)
+update_view_references(struct Select *select,
+		       int update_value, bool suppress_error)
 {
 	assert(update_value == 1 || update_value == -1);
 	struct SrcList *list = sql_select_expand_from_tables(select);
 	if (list == NULL)
-		return -1;
+		panic("Out of memory while parsing sql view request");
```

Comment:
`sql_select_expand_from_tables` should never fail with OOM since it uses `xmalloc`. Let's update its comment and code where required and drop this check.

Reply:
OK, added two prior commits that refactors `sql_select_expand_from_tables` and SQL SrcList.